service/src/main/java/com/commercetools/pspadapter/payone/SessionHandler.java

```diff
@@ -0,0 +1,202 @@
+import static java.lang.String.format;
+
+public class SessionHandler {
+
+    /**
+     * How many times to retry if {@link ConcurrentModificationException} happens.
+     */
+    private static final int RETRIES_LIMIT = 5;
+    private static final int RETRY_DELAY = 100; // msec
+    private final TenantConfig tenantConfig;
+    private final String payoneInterfaceName;
+    private final LogstashMarker tenantNameKeyValue;
+
+
+    private final CommercetoolsQueryExecutor commercetoolsQueryExecutor;
+
+    private final Logger logger;
+    private PayonePostService payonePostService;
+
+    public SessionHandler(String payoneInterfaceName, String tenantName,
+                          CommercetoolsQueryExecutor commercetoolsQueryExecutor, TenantConfig tenantConfig,
+                          PayonePostService postService) {
+        this.payoneInterfaceName = payoneInterfaceName;
+
+        this.commercetoolsQueryExecutor = commercetoolsQueryExecutor;
+
+
+        this.logger = LoggerFactory.getLogger(this.getClass());
+        tenantNameKeyValue = createTenantKeyValue(tenantName);
+        this.tenantConfig = tenantConfig;
+        this.payonePostService = postService;
+    }
+
+    /**
+     * Tries to start a payment session for the payment with the provided ID.
+     *
+     * @param paymentId identifies the payment to be processed
+     * @return the result of handling the payment
+     */
+    public PayoneResult start(@Nonnull final String paymentId) {
+        int retryCounter = 0;
+        try {
+            for (; retryCounter < RETRIES_LIMIT; retryCounter++) {
+                try {
+                    return startSession(paymentId);
+                } catch (final ConcurrentModificationException concurrentModificationException) {
+                    if (retryCounter == RETRIES_LIMIT - 1) {
+                        throw concurrentModificationException;
+                    } else {
+                        Thread.sleep(calculateVariableDelay(retryCounter));
+                    }
+                }
+            }
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            return handleConcurrentModificationException(paymentId, concurrentModificationException);
+        } catch (final NotFoundException | NoCartLikeFoundException e) {
+            return handleNotFoundException(paymentId, retryCounter, e);
+        } catch (final ErrorResponseException e) {
+            return errorResponseHandler(paymentId, retryCounter, e);
+        } catch (final Exception e) {
+            return handleException(paymentId, retryCounter, e);
+        }
+        return handleException(paymentId, retryCounter,
+                new Exception("Unknown workflow error in PaymentHandler#handlePayment"));
+    }
+
+    private PayoneResult startSession(@Nonnull final String paymentId)
+            throws ConcurrentModificationException {
```

Comment:
cannot see where `ConcurrentModificationException` is thrown....

Reply:
aa00825a2